add async router implementation

larsgrah opened the pull request:
This PR adds a way to add custom routers, and enforces an async interface for possible http requests

larsgrah committed:
add async router implementation

larsgrah committed:
remove console trace

fredlahde commented on src/Router.ts:
```diff
@@ -31,33 +34,47 @@ export class Router {
-        }
+                if (!this.componentMap.has(path)) {
+                    console.error("No component registered with the router for ", path)
+                    return reject(true);
```
return false here

larsgrah committed:
change returntype on fail to false

fredlahde commented on src/Router.ts:
```diff
@@ -71,7 +88,7 @@ export class RouterLink extends VNode {
-        if (ln.app.router.componentMap.has(ln.target)) {
+        if (ln.app.router.hasRouteRegistered(ln.target)) {
             history.pushState({}, "", document.location.pathname)
             ln.app.router.resolveRoute(ln.target);
```
Catch here and log to console.error to supress Browser generated warning about uncaught exception

larsgrah committed:
add catch to resolve route

fredlahde commented on src/Router.ts:
```diff
@@ -31,33 +34,47 @@ export class Router {
     }
 }
 
+export interface IRouter {
```
Move interfaces to top of file?

larsgrah committed:
move interfaces

fredlahde reviewed (commented):
looks good AFAIK, will test on chrome..

fredlahde committed:
Return router from useRouter, fix error handling

- Simplify resolveRoute: Remove wrapping Promise, use static functions for
resolve \ reject
- Better error message when no route is found to route to
- return router instance from useRouter / useCustomRouter

fredlahde committed:
Remove unused import

fredlahde committed:
add return types to useRouter functions

fredlahde reviewed (approved):
Works fine, I changed a few things though. Review them for yourself and then this can be merged

Good work, way more flexible now
